### PR TITLE
fix: abracadabra page

### PR DESCRIPTION
**Pull request by kirillzyusko**
### Details

The fix consist from two parts:
- preload the component in advance to be able to mount it synchronously on a demand;
- clean up the state in asynchronous way if it's a real component unmount.

Let's consider each part in more details.

#### 1️⃣ preload the component in advance to be able to mount it synchronously on a demand

If we mount the component and the component is lazy and haven't been loaded yet, then when we switch navigator we remove old navigator, but a new one navigator doesn't have routes yet, so `react-navigation` clears a state, and when new routes are attached then a default route will be selected, because previous state was lost.

So we need to mount the screen/navigator synchronously.

#### 2️⃣ clean up the state in asynchronous way if it's a real component unmount

The synchronous clean up was added here - https://github.com/Expensify/App/pull/42592

The key thing is that now `unmount` is getting fired two times. If we have an asyncronous removal, then navigation just stops working (because we perform a navigation and then instantly run unmount - unmount fires asynchronously and removes just created new state).

To fix this problem I patched `react-navigation` and added `useFullyMountedRef` hook. It's returning `ref.current=false` if component is not fully mounted (i. e. unmount is about to be fired) and will update `ref.current=true` when component finished it's mounting.

This solution is not perfect, but currently `react-navigation@6` is not fully compatible with `StrictMode`, so we have to patch it. Most likely with `react-navigation@7` this patch will not be needed, because `react-navigation@7` supporting `StrictMode`.

### Fixed Issues

$ https://github.com/Expensify/App/issues/44600
PROPOSAL: N/A

### Tests

1. Go to https://staging.new.expensify.com/
2. Log in with new Expensifail account
3. Copy the Magic link and change it to staging
4. Open a new tab and navigate to the staging magic link
5. Abracadabra page is displayed and prev tab gets redirected to inbox screen

> [!WARNING]
> Please, test a release version locally.

### Offline tests

N/A

### QA Steps

1. Go to https://staging.new.expensify.com/
2. Log in with new Expensifail account
3. Copy the Magic link and change it to staging
4. Open a new tab and navigate to the staging magic link
5. Abracadabra page is displayed and prev tab gets redirected to inbox screen

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/b6d9b913-c645-4641-b008-582ba9135060

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Comment by dangrous** [comment]
sure! Let me get that build started

**Comment by dangrous** [comment]
Do you mind double checking that that looks good @kirillzyusko and then I can send over to QA?

**Comment by kirillzyusko** [comment]
@dangrous I tested and for me it worked well in release variant 👀 I also think that potentially it may fix other issues such as https://github.com/Expensify/App/issues/48378

**Comment by dangrous** [comment]
Just requested qa's help! https://expensify.slack.com/archives/C9YU7BX5M/p1726092553187509

**Comment by dangrous** [comment]
Sounds like this passed QA, with one remaining known issue. Will get the details shortly!

**Comment by kirillzyusko** [comment]
> Sounds like this passed QA, with one remaining known issue. Will get the details shortly!

@dangrous does it mean that I can proceed with this PR further, clean it up and prepare for a review? 👀

**Comment by dangrous** [comment]
Yeah I think go ahead! I haven't heard back yet from QA on which issue still exists (I'm assuming it's not https://github.com/Expensify/App/issues/44600 which is the only one that would be concerning) but I will ping them again.

**Comment by kirillzyusko** [comment]
@aimane-chnaif @dangrous this PR is ready for a review 🎉

**Comment by kirillzyusko** [comment]
> Please merge main. 1.7k commits are behind

Done 🙌

File: package.json [hunk 1]
```diff
@@ -142,6 +142,7 @@
     "react-native": "0.75.2",
     "react-native-android-location-enabler": "^2.0.1",
     "react-native-blob-util": "0.19.4",
+    "react-native-bundle-splitter": "^3.0.1",
```
[the remark this answers is not in the document]
Comment (kirillzyusko) [reply package.json:136]: Submitted an issue here: https://github.com/Expensify/App/issues/49519

**Comment by dangrous** [comment]
Hey @kirillzyusko finally got the response from QA - looks like the PR still reproduces these issues:
https://github.com/Expensify/App/issues/48378
https://github.com/Expensify/App/issues/48960

Any chance you could take a look at those?

**Comment by kirillzyusko** [comment]
@dangrous so we want to fix all of these issues in a single PR? Or do you want me to look and discover a root problem for these bugs?

**Comment by kirillzyusko** [comment]
> https://github.com/Expensify/App/issues/48378

I can confirm this is still reproducible in this branch.

> https://github.com/Expensify/App/issues/48960

I can not reproduce it from this branch 🤷‍♂️

**Comment by dangrous** [comment]
Okay so [one of those](https://github.com/Expensify/App/issues/48960) was closed and QA suggested this PR would reopen it. That one we should definitely make sure we don't reintroduce with this.

The other one was open, I was going to say we don't have to handle it here BUT they just [closed](https://github.com/Expensify/App/issues/48378#issuecomment-2352312029) it in favor of this PR haha. It sounds like it might not be reproducible anymore.

So TLDR - let's see if either of those occur with this PR, and we only need to fix if they still do.

**Comment by kirillzyusko** [comment]
> Okay so https://github.com/Expensify/App/issues/48960 was closed and QA suggested this PR would reopen it. That one we should definitely make sure we don't reintroduce with this.

Well, I can rebase this PR to latest main because right now it's ~1000 commits behind. I couldn't reproduce the problem, but I hardly believe this PR can introduce such behavior 😅

**Comment by dangrous** [comment]
Confirmed, I don't see that happening either. I think we're okay?

**Comment by dangrous** [comment]
As for https://github.com/Expensify/App/issues/48378 - you said you could reproduce that one here? Let's see if we can fix it too, since even if we don't show the inbox (this issue) we probably should make sure we show the correct (non-expired) page, too. Does that make sense?

**Comment by garrettmknight** [comment]
@kirillzyusko any update here?

**Comment by dangrous** [comment]
It looks actually like it's being fixed separately so yeah, you can ignore it here!

**Comment by kirillzyusko** [comment]
> It looks actually like it's being fixed separately so yeah, you can ignore it here!

Cool! So we can review this PR then, right? 👀

**Comment by dangrous** [comment]
Yep! though we're waiting on approval of that library right?

**Comment by kirillzyusko** [comment]
> Yep! though we're waiting on approval of that library right?

Oh, yeah, right!

**Comment by kirillzyusko** [comment]
Closing PR because the behavior was re-defined: https://github.com/Expensify/App/issues/44600#issuecomment-2460068450
